modules/core/src/smithy4s/codecs/FieldSkipCompiler.scala

```diff
@@ -0,0 +1,144 @@
+      else compileOptional(field)
+
+    }
+
+    def compileOptional[A](
+        field: Field[?, A]
+    ): FieldSkipCompiler.ShouldSkip[A]
+
+  }
+
+  case object NeverSkip extends FieldSkipCompiler {
+    def compile[A](field: Field[_, A]): ShouldSkip[A] = Function.const(false)
+  }
+
+  private def asEmptyCollectionPredicate[F[_], A](
+      schema: Schema[A]
+  ): Option[A => Boolean] = {
+    import Schema._
+    schema match {
+      case c: CollectionSchema[f, A @unchecked] =>
+        Some(collectionA => c.tag.isEmpty(collectionA.asInstanceOf[f[A]]))
+      case b: BijectionSchema[inner, A @unchecked] =>
+        asEmptyCollectionPredicate[F, inner](b.underlying).map(predicateInner =>
+          collectionA => predicateInner(b.bijection.from(collectionA))
+        )
+      case r: RefinementSchema[inner, A @unchecked] =>
+        asEmptyCollectionPredicate[F, inner](r.underlying).map(predicateInner =>
+          collectionA => predicateInner(r.refinement.from(collectionA))
+        )
+      case o: OptionSchema[inner] =>
+        asEmptyCollectionPredicate(o.underlying)
+          .map(predicateInner =>
+            collectionA =>
+              collectionA.asInstanceOf[Option[inner]].exists(predicateInner)
+          )
+      case _: MapSchema[k, v] =>
+        Some(collectionA => collectionA.asInstanceOf[Map[k, v]].isEmpty)
+      case LazySchema(suspend)     => asEmptyCollectionPredicate(suspend.value)
+      case _: EnumerationSchema[_] => None
+      case _: StructSchema[_]      => None
+      case _: UnionSchema[_]       => None
+      case _: PrimitiveSchema[_]   => None
+    }
+  }
+
+  private case object skipIfEmptyOptionalCollection
+      extends FieldSkipCompiler.SkipNonRequired {
+
+    def compileOptional[A](field: Field[?, A]): ShouldSkip[A] = {
+      asEmptyCollectionPredicate(field.schema) match {
+        case None          => Function.const(false)
+        case Some(isEmpty) => isEmpty
+      }
+    }
+  }
+
+  val SkipIfEmptyOptionalCollection: FieldSkipCompiler =
+    skipIfEmptyOptionalCollection
+
+  case object SkipIfEmptyCollection extends FieldSkipCompiler {
+
+    def compile[A](field: Field[_, A]): ShouldSkip[A] = {
+      asEmptyCollectionPredicate(field.schema) match {
+        case None          => Function.const(false)
+        case Some(isEmpty) => isEmpty
+      }
+    }
+  }
+
+  private case object skipIfDefaultOptionals
+      extends FieldSkipCompiler.SkipNonRequired {
+    def compileOptional[A](field: Field[?, A]): ShouldSkip[A] = {
+      // Optional fields have None as their default, so we need to make sure not to skip them here
+      a => a != None && field.isDefaultValue(a)
+    }
+  }
+
+  val SkipIfDefaultOptionals: FieldSkipCompiler = skipIfDefaultOptionals
+
+  private case object skipIfEmptyOptionals
+      extends FieldSkipCompiler.SkipNonRequired {
+    def compileOptional[A](field: Field[?, A]): ShouldSkip[A] = { a =>
+      a == None
```

Comment:
this relates to my previous comment about optional, but I feel icky about comparing an `A` that we don't know anything about to a None.

And now I just remembered about the Dynamic module - that's one space in which `==` can not always be relied upon (e.g. when comparing structs, `==` would just compare arrays (`smithy4s.dynamic.internals.DynStruct`) which is not what you want). Thankfully `Option` is a thing in Dynamic, but in theory it might be replaceable with a null as an optimization...

so I'd at least make sure to have a couple tests for these which go through `DSI`.

Reply:
`DSI` is used in `AwsComplianceSuite` so I think that it is covered already. Though I am not opposed to adding more tests for it, just not sure where to put them.